src/lib/index/wb_cache.cpp

```diff
@@ -155,48 +142,206 @@ void IndexWBCache::read_buf(bnodeid_t id, BtreeNodePtr& node, node_initializer_t
+                                 IndexBufferPtrList const& freed_node_bufs, CPContext* cp_ctx) {
+    IndexCPContext* icp_ctx = r_cast< IndexCPContext* >(cp_ctx);
+    if (parent_buf) { link_buf(parent_buf, child_buf, false /* is_sibling_link */, cp_ctx); }
 
-        chain->m_next_buffer = second;
-        second->m_wait_for_leaders.increment(1);
+#ifdef _PRERELEASE
+    set_crash_flips(parent_buf, child_buf, new_node_bufs, freed_node_bufs);
+#endif
+
+    for (auto const& buf : new_node_bufs) {
+        link_buf(child_buf, buf, true /* is_sibling_link */, cp_ctx);
+    }
+
+    for (auto const& buf : freed_node_bufs) {
+        if (!buf->m_wait_for_down_buffers.testz()) {
+            // This buffer has some down bufs depending on it. It can happen for an upper level interior node, where
+            // lower level node (say leaf) has split causing it to write entries in this node, but this node is now
+            // merging with other node, causing it to free. In these rare instances, we link this node to the new
+            // node resulting in waiting for all the down bufs to be flushed before up buf can flush (this buf is
+            // not written anyways)
+            link_buf(child_buf, buf, true /* is_sibling_link */, cp_ctx);
+        }
     }
 
-    return {second_copied, third_copied};
+    icp_ctx->add_to_txn_journal(index_ordinal, child_buf->m_up_buffer,
+                                new_node_bufs.size() ? new_node_bufs[0]->m_up_buffer : nullptr, new_node_bufs,
+                                freed_node_bufs);
 }
 
-void IndexWBCache::prepend_to_chain(const IndexBufferPtr& first, const IndexBufferPtr& second) {
-    assert(first->m_next_buffer.lock() != second);
-    assert(first->m_next_buffer.lock() == nullptr);
-    first->m_next_buffer = second;
-    second->m_wait_for_leaders.increment(1);
+void IndexWBCache::link_buf(IndexBufferPtr const& up_buf, IndexBufferPtr const& down_buf, bool is_sibling_link,
+                            CPContext* cp_ctx) {
+    HS_DBG_ASSERT_NE((void*)up_buf->m_up_buffer.get(), (void*)down_buf.get(), "Cyclic dependency detected");
+    IndexBufferPtr real_up_buf = up_buf;
+    IndexCPContext* icp_ctx = r_cast< IndexCPContext* >(cp_ctx);
+
+    // Condition 1: If the down buffer and up buffer are both created by the current cp_id, unconditionally we need
+    // to link it with up_buffer's up_buffer. In other words, there should never a link between down and up buffers
+    // created in current generation (cp). In real terms, it means all new buffers can be flushed independently to
+    // each other and dependency is needed only for the buffers created in previous cps.
+    if ((down_buf->m_created_cp_id == icp_ctx->id()) && (up_buf->m_created_cp_id == icp_ctx->id())) {
+        real_up_buf = up_buf->m_up_buffer;
+        HS_DBG_ASSERT(real_up_buf,
+                      "Up buffer is newly created in this cp, but it doesn't have its own up_buffer, its not expected");
+    }
```

Comment:
Note for me: is it a correct assert here ? `up_buf->m_created_cp_id == icp_ctx->id() && down_buf->dirtied_cp_id < up_buf->m_created_cp_id`